File: web/src/types/task.ts
import type { Timestamp } from "firebase/firestore";

/**
 * Mirrors the TypeScript representation in /docs/data-model.md exactly.
 * Do not diverge from that document without updating it first — it is the
 * canonical schema shared with the Android client.
 */
export type Priority = "high" | "medium" | "low";

export interface Task {
  id: string;
  title: string;
  notes: string | null;
  completed: boolean;
  completedAt: Timestamp | null;
  dueDate: Timestamp | null;
  priority: Priority;
  order: number;
  createdAt: Timestamp | null;
  updatedAt: Timestamp | null;
  createdByUid: string;
  updatedByUid: string;
}

/**
 * Shape of a new task as created from the UI, before Firestore assigns
 * server-generated fields (id, createdAt, updatedAt, createdByUid, updatedByUid).
 */
export interface NewTaskInput {
  title: string;
  notes: string | null;
  dueDate: Date | null;
  priority: Priority;
  order: number;
}

/**
 * Partial patch applied to an existing task. updatedAt/updatedByUid are
 * always stamped by the repository layer, not the caller.
 */
export interface TaskPatch {
  title?: string;
  notes?: string | null;
  dueDate?: Date | null;
  priority?: Priority;
  completed?: boolean;
}

File: android/app/src/main/java/com/wikzo/todo/data/model/Task.kt
package com.wikzo.todo.data.model

import com.google.firebase.Timestamp

/**
 * Mirrors `syncGroups/{groupId}/tasks/{taskId}` as documented in /docs/data-model.md.
 * Keep this class in lockstep with that doc's "Kotlin representation" section --
 * Firestore has no server-side schema enforcement, so this is where drift would show up.
 */
data class Task(
    val id: String = "",
    val title: String = "",
    val notes: String? = null,
    val completed: Boolean = false,
    val completedAt: Timestamp? = null,
    val dueDate: Timestamp? = null,
    val priority: Priority = Priority.MEDIUM,
    val order: Double = 0.0,
    val createdAt: Timestamp? = null,
    val updatedAt: Timestamp? = null,
    val createdByUid: String = "",
    val updatedByUid: String = "",
)

/**
 * Firestore stores priority as a closed lowercase string enum ("high" | "medium" | "low"),
 * never a number, so the wire format stays self-describing. [firestoreValue] /
 * [fromFirestoreValue] are the single place that mapping happens.
 *
 * Declaration order (HIGH, MEDIUM, LOW) doubles as sort order -- see
 * TaskListViewModel's comparator, which sorts on `priority.ordinal`.
 */
enum class Priority {
    HIGH,
    MEDIUM,
    LOW;

    val firestoreValue: String
        get() = when (this) {
            HIGH -> "high"
            MEDIUM -> "medium"
            LOW -> "low"
        }

    companion object {
        fun fromFirestoreValue(value: String?): Priority = when (value) {
            "high" -> HIGH
            "medium" -> MEDIUM
            "low" -> LOW
            else -> MEDIUM
        }
    }
}
